lib/listen/adapter/darwin.rb

```diff
@@ -35,59 +35,45 @@ def self.usable?
 
       private
 
-      # NOTE: each directory gets a DIFFERENT callback!
       def _configure(dir, &callback)
-        require 'rb-fsevent'
-        opts = { latency: options.latency }
-
-        @workers ||= ::Queue.new
-        @workers << FSEvent.new.tap do |worker|
-          _log :debug, "fsevent: watching: #{dir.to_s.inspect}"
-          worker.watch(dir.to_s, opts, &callback)
-        end
+        @callbacks[dir] = callback
       end
 
       def _run
-        first = @workers.pop
-
-        # NOTE: _run is called within a thread, so run every other
-        # worker in it's own thread
-        _run_workers_in_background(_to_array(@workers))
-        _run_worker(first)
+        require 'rb-fsevent'
+        worker = FSEvent.new
+        dirs_to_watch = @callbacks.keys.map(&:to_s)
+        _log(:info) { "fsevent: watching: #{dirs_to_watch.inspect}" }
+        worker.watch(dirs_to_watch, { latency: options.latency }, &method(:_process_changes))
+        Listen::Internals::ThreadPool.add { _run_worker(worker) }
       end
 
-      def _process_event(dir, event)
-        _log :debug, "fsevent: processing event: #{event.inspect}"
-        event.each do |path|
-          new_path = Pathname.new(path.sub(%r{\/$}, ''))
-          _log :debug, "fsevent: #{new_path}"
-          # TODO: does this preserve symlinks?
-          rel_path = new_path.relative_path_from(dir).to_s
-          _queue_change(:dir, dir, rel_path, recursive: true)
+      def _process_changes(dirs)
+        dirs.each do |dir|
```

Comment:
I'm aware there are two loops here but for the average use case, this is going to consist of only a couple of directories that change each time.